Drop use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in StyleGradient.cpp

cdumez opened the pull request:
#### 7134ab8e431f7d0d5caa3e500b59827cebb2b49d
<pre>
Drop use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in StyleGradient.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=286227">https://bugs.webkit.org/show_bug.cgi?id=286227</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc&gt;::insertSpan):
(WTF::Malloc&gt;::insertVector):
* Source/WebCore/style/values/images/StyleGradient.cpp:
(WebCore::Style::computeStops):

Canonical link: <a href="https://commits.webkit.org/289167@main">https://commits.webkit.org/289167@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef950a547f18e9edfed30533cf165f3e5a2206b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90673 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36586 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87634 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5477 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13258 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66514 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24325 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88596 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77710 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46797 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4077 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31973 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35656 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/78594 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74744 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32829 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92335 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84584 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12905 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9468 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/75236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13119 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73543 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74373 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18600 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17042 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/5022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13338 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12924 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18305 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106974 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12723 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25783 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16156 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14505 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/values/images/StyleGradient.cpp:
```diff
@@ -518,7 +516,7 @@ template<typename GradientAdapter, typename StyleGradient> GradientColorStops co
 
         stops.remove(x);
-        stops.insert(x, newStops, 9);
+        stops.insertSpan(x, std::span<const ResolvedGradientStop, 9> { newStops });
```
I'm not sure it is an important optimization to make `9` a compile time constant here. What happens if you just construct `std::span` without providing template parameters?

weinig replied:
It should deduce it anyway - https://en.cppreference.com/w/cpp/container/span/deduction_guides (#4)

cdumez replied:
I tried `std::span { }` first and it didn't build. The issue is that it deduces the wrong constness. I would not have written it out unless the compiler had forced me to :)

cdumez replied:
Yes, it's not building without the explicit template parameters:
```
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource443.cpp:8:
/Volumes/Work/WebKit/OpenSource/Source/WebCore/style/values/images/StyleGradient.cpp:519:15: error: no matching member function for call to 'insertSpan'
  519 |         stops.insertSpan(x, std::span { newStops });
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource443.cpp:1:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebCore/WebCorePrefix.h:168:
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/text/AtomString.h:24:
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/text/AtomStringImpl.h:24:
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/text/UniquedStringImpl.h:28:
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/text/StringImpl.h:39:
/Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/Vector.h:925:73: note: candidate template ignored: cannot deduce a type for 'U' that would make 'const U' equal 'WebCore::Style::ResolvedGradientStop'
  925 |     template<typename U, std::size_t Extent = std::dynamic_extent> void insertSpan(size_t position, std::span<const U, Extent>);
1 error generated.
```

cdumez replied:
Ok, looks like I was able to make it work by updating the definition of insertSpan() not not require a span to `const`.

geoffreygaren commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1623,21 +1623,21 @@ inline void Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::app
+template<typename U, std::size_t Extent>
+void Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::insertSpan(size_t position, std::span<U, Extent> data)
 {
     ASSERT_WITH_SECURITY_IMPLICATION(position <= size());
```
Not new to this patch, but we seem to be missing a release_assert that position is inbounds here.

cdumez replied:
I'll follow up after some A/B testing.

cdumez committed:
Drop use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in StyleGradient.cpp
https://bugs.webkit.org/show_bug.cgi?id=286227

Reviewed by Darin Adler.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc>::insertSpan):
(WTF::Malloc>::insertVector):
* Source/WebCore/style/values/images/StyleGradient.cpp:
(WebCore::Style::computeStops):

Canonical link: https://commits.webkit.org/289167@main
